base_local_plannerのパラメータを見直す

今後のパラメータチューニングが円滑に進むように，すべてのパラメータをROS Wikiに準じた順番で書き下す．

とりあえず，`TrajectoryPlannerROS`と`DWAPlannerROS`についてのパラメータファイルを作る．

現在設定していない他のパラメータをデフォルト値のまま`trajectory_planner.yml`に追記し，Gazebo上でナビゲーションのテストを行った結果，変更前では動作していたものが正常に動作しなくなった．

具体的には，適当な場所にゴールを設定したときにその場から動けずに回転運動を行うのみになった．

ナビゲーションパッケージはエラーを出力し続けた．エラーメッセージは以下のようなものが連続で流れてきた．

```console
[ WARN] [1597979578.303834772, 153.544000000]: Could not transform the global plan to the frame of the controller
[ERROR] [1597979578.906482459, 154.145000000]: Extrapolation Error: Lookup would require extrapolation into the future.  Requested time 154.129000000 but the latest data is at time 154.104000000, when looking up transform from frame [odom] to frame [map]
```

TF関係のエラーが吐かれている．原因の解明を急ぐ．

--- a/.gitignore
+++ b/.gitignore
@@ -51,3 +51,4 @@ qtcreator-*
 CATKIN_IGNORE
 
 .vscode/
+*swp

--- a/adamr2_bringup/config/f310.config.yml
+++ b/adamr2_bringup/config/f310.config.yml
@@ -1,10 +1,15 @@
 axis_linear: 1
-scale_linear: 0.2
-scale_linear_turbo: 0.5
+#scale_linear: 0.2
+#scale_linear_turbo: 0.5
+scale_linear: 0.8
+scale_linear_turbo: 1.2
 
 axis_angular: 0
-scale_angular: 0.3925
-scale_angular_turbo: 0.785
+#scale_angular: 0.3925
+#scale_angular_turbo: 0.785
+
+scale_angular: 1.57
+scale_angular_turbl: 3.14
 
 enable_button: 1
 enable_turbo_button: 2

--- a/adamr2_navigation/config/amcl/amcl_diff.yml
+++ b/adamr2_navigation/config/amcl/amcl_diff.yml
@@ -1,11 +1,11 @@
-min_particles: 100  # default = 100
-max_particles: 5000 # default = 5000
+min_particles: 30  # default = 100
+max_particles: 100 # default = 5000
 
 kld_err: 0.01 # default = 0.01
 kld_z: 0.99   # default = 0.99
 
-update_min_d: 0.2 # default = 0.2
-update_min_a: 0.5 # default = PI / 6.0 ~= 0.5
+update_min_d: 0.05 # default = 0.2
+update_min_a: 0.08 # default = PI / 6.0 ~= 0.5
 
 resample_interval: 1      #  default = 2
 transform_tolerance: 0.1  # default = 0.1

--- a/adamr2_navigation/config/amcl/amcl_diff.yml
+++ b/adamr2_navigation/config/amcl/amcl_diff.yml
@@ -5,7 +5,7 @@ kld_err: 0.01 # default = 0.01
 kld_z: 0.99   # default = 0.99
 
 update_min_d: 0.05 # default = 0.2
-update_min_a: 0.08 # default = PI / 6.0 ~= 0.5
+update_min_a: 0.05 # default = PI / 6.0 ~= 0.5
 
 resample_interval: 1      #  default = 2
 transform_tolerance: 0.1  # default = 0.1

--- a/adamr2_navigation/config/planner/local/dwa_local_planner.yml
+++ b/adamr2_navigation/config/planner/local/dwa_local_planner.yml
@@ -1,0 +1,45 @@
+base_local_planner: dwa_local_planner/DWAPlannerROS
+DWAPlannerROS:
+  # Robot Configuration Parameters
+  acc_lim_x: 1.0
+  acc_lim_y: 0.0
+  #acc_lim_trans: 1.0
+  acc_lim_theta: 3.0
+
+  max_vel_trans: 1.0
+  min_vel_trans: 0.0
+  max_vel_x: 1.0
+  min_vel_x: 0.0
+  max_vel_y: 0.0
+  min_vel_y: 0.0
+  max_vel_theta: 1.57
+  min_vel_theta: 0.0
+
+  # Goal Tolerance Parameters
+  yaw_goal_tolerance: 0.25
+  xy_goal_tolerance: 0.25
+  latch_xy_goal_tolerance: false
+
+  # Forward Simulation Parameters
+  sim_time: 4.0
+  sim_granularity: 0.050
+  vx_samples: 20
+  vy_samples: 1
+  vth_samples: 40
+  controller_frequency: 10.0
+
+  # Trajectory Scoring Parameters
+  #path_distance_bias: 32.0
+  #goal_distance_bias: 24.0
+  #occdist_scale: 0.01
+  #forward_point_distance: 0.325
+  #stop_time_buffer: 0.2
+  #scaling_speed: 1.2
+  #max_scaling_factor: 0.2
+  #publish_cost_grid: true
+
+  # Oscillation Preventation Parameters
+  #oscillation_reset_dist: 0.3
+
+  # Global Plan Parameters
+  #prune_plan: true

--- a/adamr2_navigation/config/planner/local/trajectory_planner.yml
+++ b/adamr2_navigation/config/planner/local/trajectory_planner.yml
@@ -5,11 +5,11 @@ TrajectoryPlannerROS:
   acc_lim_theta: 1.0    # default: 3.2
 
   max_vel_x: 0.8        # default: 0.5
-  min_vel_x: -0.8       # default: 0.1
+  min_vel_x: 0.2       # default: 0.1
   max_vel_theta: 1.0    # default: 1.0
   min_vel_theta: -1.0   # default: -1.0
   min_in_place_vel_theta: 1.0 # default: 0.4
-  escape_vel: -0.1      # default: -0.1
+  escape_vel: -0.5      # default: -0.1
 
   holonomic_robot: false  # default: true
 
@@ -19,20 +19,20 @@ TrajectoryPlannerROS:
   latch_xy_goal_tolerance: true # default: false
 
   # Forward Simulation Parameters
-  sim_time: 1.0                     # default: 1.0
-  sim_granularity: 0.025            # default: 0.025
-  angular_sim_granularity: 0.025    # default: 0.025
-  vx_samples: 3                     # default: 3
-  vtheta_samples: 20               # default: 20
-  controller_frequency: 20.0        # default: 20.0
+  sim_time: 4.0                     # default: 1.0
+  sim_granularity: 0.050            # default: 0.025
+  angular_sim_granularity: 0.050    # default: 0.025
+  vx_samples: 20                    # default: 3
+  vtheta_samples: 40                # default: 20
+  controller_frequency: 10.0        # default: 20.0
 
   # Trajectory Scoring Parameters
-  meter_scoring: false          # default: false
-  path_distance_bias: 0.6              # default: 0.6
-  goal_distance_bias: 0.8              # default: 0.8
+  meter_scoring: true           # default: false
+  path_distance_bias: 0.6       # default: 0.6
+  goal_distance_bias: 0.8       # default: 0.8
   occdist_scale: 0.01           # default: 0.01
   heading_lookahead: 0.325      # default: 0.325
-  heading_scoring: false        # default: false
+  heading_scoring: true         # default: false
   heading_scoring_timestep: 0.8 # default: 0.8
   dwa: true                     # default: true
   publish_cost_grid_pc: false   # default: false

--- a/adamr2_navigation/config/planner/local/trajectory_planner.yml
+++ b/adamr2_navigation/config/planner/local/trajectory_planner.yml
@@ -28,8 +28,8 @@ TrajectoryPlannerROS:
 
   # Trajectory Scoring Parameters
   meter_scoring: false          # default: false
-  pdist_scale: 0.6              # default: 0.6
-  gdist_scale: 0.8              # default: 0.8
+  path_distance_bias: 0.6              # default: 0.6
+  goal_distance_bias: 0.8              # default: 0.8
   occdist_scale: 0.01           # default: 0.01
   heading_lookahead: 0.325      # default: 0.325
   heading_scoring: false        # default: false

--- a/adamr2_navigation/config/planner/local/trajectory_planner.yml
+++ b/adamr2_navigation/config/planner/local/trajectory_planner.yml
@@ -1,14 +1,45 @@
 base_local_planner: base_local_planner/TrajectoryPlannerROS
 TrajectoryPlannerROS:
-  max_vel_x: 0.8
-  min_vel_x: -0.8
-  max_vel_theta: 1.0
-  min_vel_theta: -1.0
-  min_in_place_vel_theta: 1.0
-  acc_lim_x: 0.8
-  acc_lim_theta: 1.0
+  # Robot Configuration Parameters
+  acc_lim_x: 0.8        # default: 2.5
+  acc_lim_theta: 1.0    # default: 3.2
 
-  holonomic_robot: false
+  max_vel_x: 0.8        # default: 0.5
+  min_vel_x: -0.8       # default: 0.1
+  max_vel_theta: 1.0    # default: 1.0
+  min_vel_theta: -1.0   # default: -1.0
+  min_in_place_vel_theta: 1.0 # default: 0.4
+  escape_vel: -0.1      # default: -0.1
 
-  yaw_goal_tolerance: 0.25
-  xy_goal_tolerance: 0.25
+  holonomic_robot: false  # default: true
+
+  # Goal Tolerance Parameters
+  yaw_goal_tolerance: 0.25  # default: 0.05
+  xy_goal_tolerance: 0.25   # default: 0.10
+  latch_xy_goal_tolerance: true # default: false
+
+  # Forward Simulation Parameters
+  sim_time: 1.0                     # default: 1.0
+  sim_granularity: 0.025            # default: 0.025
+  angular_sim_granularity: 0.025    # default: 0.025
+  vx_samples: 3                     # default: 3
+  vtheta_samplesp: 20               # default: 20
+  controller_frequency: 20.0        # default: 20.0
+
+  # Trajectory Scoring Parameters
+  meter_scoring: false          # default: false
+  pdist_scale: 0.6              # default: 0.6
+  gdist_scale: 0.8              # default: 0.8
+  occdist_scale: 0.01           # default: 0.01
+  heading_lookahead: 0.325      # default: 0.325
+  heading_scoring: false        # default: false
+  heading_scoring_timestep: 0.8 # default: 0.8
+  dwa: true                     # default: true
+  publish_cost_grid_pc: false   # default: false
+  global_frame_id: odom         # default: odom
+
+  # Oscillation Prevention Parameters
+  oscillation_reset_dist: 0.05  # default: 0.05
+
+  # Global Plan Parameters
+  prune_plan: true  # default: true

--- a/adamr2_navigation/config/planner/local/trajectory_planner.yml
+++ b/adamr2_navigation/config/planner/local/trajectory_planner.yml
@@ -23,7 +23,7 @@ TrajectoryPlannerROS:
   sim_granularity: 0.025            # default: 0.025
   angular_sim_granularity: 0.025    # default: 0.025
   vx_samples: 3                     # default: 3
-  vtheta_samplesp: 20               # default: 20
+  vtheta_samples: 20               # default: 20
   controller_frequency: 20.0        # default: 20.0
 
   # Trajectory Scoring Parameters

--- a/adamr2_navigation/launch/adamr2_navigation.launch
+++ b/adamr2_navigation/launch/adamr2_navigation.launch
@@ -17,16 +17,15 @@
     <!-- BASE GLOBAL PLANNER -->
     <rosparam file="$(find adamr2_navigation)/config/planner/global/navfn_planner.yml" command="load" />
     <!-- BASE LOCAL PLANNER -->
-    <rosparam file="$(find adamr2_navigation)/config/planner/local/trajectory_planner.yml" command="load" />
-    <!-- RECOVERY -->
-    <rosparam file="$(find adamr2_navigation)/config/planner/recovery_behaviors.yml" command="load"/>
+    <rosparam file="$(find adamr2_navigation)/config/planner/local/dwa_local_planner.yml" command="load" />
 
     <!-- MOVE BASE -->
-    <param name="controller_frequency" value="5.0" />
-    <param name="controller_patience" value="15.0" />
+    <param name="controller_frequency" value="10.0" />
+    <!--<param name="controller_patience" value="15.0" />
     <param name="max_planning_retries" value="10" />
     <param name="oscillation_timeout" value="10.0" />
     <param name="clearing_rotation_allowed" value="true" />
+    <param name="recovery_behavior_enabled" value="false"/>-->
 
     <!-- remap cmd_vel topic for gazebo -->
     <remap from="cmd_vel" to="/adamr2/diff_drive_controller/cmd_vel"/>

--- a/adamr2_navigation/launch/amcl.launch
+++ b/adamr2_navigation/launch/amcl.launch
@@ -1,0 +1,14 @@
+<launch>
+  <!-- MAP SERVER -->
+  <arg name="map_file" default="$(env HOME)/experiment_map/large_map/map.yaml"/>
+  <node name="map_server" pkg="map_server" type="map_server" args="$(arg map_file)">
+    <param name="frame_id" value="map" />
+  </node>
+
+  <!-- amcl -->
+  <node pkg="amcl" type="amcl" name="amcl" output="screen">
+    <rosparam file="$(find adamr2_navigation)/config/amcl/amcl_diff.yml" command="load" />
+  </node>
+
+  <node name="rviz" pkg="rviz" type="rviz" required="true" args="-d $(find adamr2_navigation)/rviz/navigation.rviz" />
+</launch>

--- a/adamr2_navigation/rviz/navigation.rviz
+++ b/adamr2_navigation/rviz/navigation.rviz
@@ -6,6 +6,7 @@ Panels:
       Expanded:
         - /Global Options1
         - /Status1
+        - /PoseArray1
       Splitter Ratio: 0.5
     Tree Height: 728
   - Class: rviz/Selection
@@ -223,6 +224,83 @@ Visualization Manager:
                   {}
       Update Interval: 0
       Value: true
+    - Alpha: 1
+      Buffer Length: 1
+      Class: rviz/Path
+      Color: 25; 255; 0
+      Enabled: true
+      Head Diameter: 0.30000001192092896
+      Head Length: 0.20000000298023224
+      Length: 0.30000001192092896
+      Line Style: Lines
+      Line Width: 0.029999999329447746
+      Name: Path
+      Offset:
+        X: 0
+        Y: 0
+        Z: 0
+      Pose Color: 255; 85; 255
+      Pose Style: None
+      Radius: 0.029999999329447746
+      Shaft Diameter: 0.10000000149011612
+      Shaft Length: 0.10000000149011612
+      Topic: /move_base/NavfnROS/plan
+      Unreliable: false
+      Value: true
+    - Alpha: 1
+      Buffer Length: 1
+      Class: rviz/Path
+      Color: 0; 0; 255
+      Enabled: true
+      Head Diameter: 0.30000001192092896
+      Head Length: 0.20000000298023224
+      Length: 0.30000001192092896
+      Line Style: Lines
+      Line Width: 0.029999999329447746
+      Name: Path
+      Offset:
+        X: 0
+        Y: 0
+        Z: 0
+      Pose Color: 255; 85; 255
+      Pose Style: None
+      Radius: 0.029999999329447746
+      Shaft Diameter: 0.10000000149011612
+      Shaft Length: 0.10000000149011612
+      Topic: /move_base/TrajectoryPlannerROS/local_plan
+      Unreliable: false
+      Value: true
+    - Alpha: 1
+      Arrow Length: 0.30000001192092896
+      Axes Length: 0.30000001192092896
+      Axes Radius: 0.009999999776482582
+      Class: rviz/PoseArray
+      Color: 255; 25; 0
+      Enabled: true
+      Head Length: 0.07000000029802322
+      Head Radius: 0.029999999329447746
+      Name: PoseArray
+      Shaft Length: 0.23000000417232513
+      Shaft Radius: 0.009999999776482582
+      Shape: Arrow (Flat)
+      Topic: /particlecloud
+      Unreliable: false
+      Value: true
+    - Alpha: 1
+      Axes Length: 1
+      Axes Radius: 0.10000000149011612
+      Class: rviz/Pose
+      Color: 255; 25; 0
+      Enabled: true
+      Head Length: 0.30000001192092896
+      Head Radius: 0.10000000149011612
+      Name: Pose
+      Shaft Length: 1
+      Shaft Radius: 0.05000000074505806
+      Shape: Arrow
+      Topic: /move_base_simple/goal
+      Unreliable: false
+      Value: true
   Enabled: true
   Global Options:
     Background Color: 48; 48; 48
@@ -251,25 +329,25 @@ Visualization Manager:
   Views:
     Current:
       Class: rviz/Orbit
-      Distance: 50.54310607910156
+      Distance: 6.3158721923828125
       Enable Stereo Rendering:
         Stereo Eye Separation: 0.05999999865889549
         Stereo Focal Distance: 1
         Swap Stereo Eyes: false
         Value: false
       Focal Point:
-        X: -3.7415454387664795
-        Y: -1.8285777568817139
-        Z: 1.0258747339248657
+        X: -2.0138967037200928
+        Y: -1.7549645900726318
+        Z: 1.6115680932998657
       Focal Shape Fixed Size: true
       Focal Shape Size: 0.05000000074505806
       Invert Z Axis: false
       Name: Current View
       Near Clip Distance: 0.009999999776482582
-      Pitch: 1.5697963237762451
+      Pitch: 0.9747968912124634
       Target Frame: <Fixed Frame>
       Value: Orbit (rviz)
-      Yaw: 6.12857723236084
+      Yaw: 0.015393390320241451
     Saved: ~
 Window Geometry:
   Displays:
@@ -287,5 +365,5 @@ Window Geometry:
   Views:
     collapsed: true
   Width: 927
-  X: 67
+  X: 993
   Y: 27

